src/openforms/submissions/form_logic.py

```diff
@@ -182,27 +184,26 @@ def evaluate_form_logic(
         # debounced client-side data changes
         data_diff = {}
 
-        for key, new_value in step.data.items():
-            original_value = initial_data.get(key)
+        for key in map(lambda x: x["key"], configuration["components"]):
+            new_value = updated_step_data.get(key, empty)
+            original_value = initial_data.get(key, empty)
             # Reset the value of any field that may have become hidden again after evaluating the logic
-            if original_value:
+            if original_value is not empty:
                 component = get_component(configuration, key)
-                default = get_component_default_value(component)
                 if (
                     component
                     and component.get("hidden")
                     and component.get("clearOnHide")
                 ):
-                    data_diff[key] = default or ""
+                    data_diff[key] = ""
                     continue
 
-            if new_value == original_value:
+            if new_value is empty or new_value == original_value:
                 continue
             data_diff[key] = new_value
 
         # only return the 'overrides'
         step.data = DirtyData(data_diff)
-
```

Comment:
this line is here for a visual separation with nasty instance-state mutations, please put it back.